src/Language/PureScript/Infernal.hs

```diff
@@ -0,0 +1,133 @@
+import Language.PureScript.Types as P
+import Language.PureScript.TypeChecker.Skolems as Skolem
+import Language.PureScript.TypeChecker.Synonyms as P
+import Language.PureScript.ModuleDependencies as P
+
+-- REPL functions
+loadModules :: [Text] -> IO P.Environment
+loadModules = loadEnv . map (\m -> "./externs/" <> toS m <> ".json")
+
+loadEnv :: MonadIO m => [FilePath] -> m P.Environment
+loadEnv modules = do
+  mkEnv <$> traverse readExternFile modules
+  where
+    readExternFile :: MonadIO m => FilePath -> m P.ExternsFile
+    readExternFile fp = do
+      parseResult <- liftIO (decodeStrict <$> BS.readFile fp)
+      case parseResult of
+        Nothing -> internalError "parsing an extern failed"
+        Just externs -> pure externs
+
+-- Potentially needed for psc-ide integration
+
+mkEnv :: [P.ExternsFile] -> P.Environment
+mkEnv externs =
+  let externsMap = foldr (\e -> Map.insert (P.efModuleName e) e) Map.empty externs
+  in foldl' (flip P.applyExternsFileToEnvironment) P.initEnvironment (sortExterns externsMap)
+
+sortExterns :: Map P.ModuleName P.ExternsFile -> [P.ExternsFile]
+sortExterns externs = do
+  sorted' <- runExceptT . P.sortModules . map mkShallowModule . Map.elems $ externs
+  case sorted' of
+    Left _ -> internalError "There was a cycle in the dependencies"
+    Right (sorted, _) -> do
+      mapMaybe getExtern ((P.efModuleName <$> Map.elems externs) `inOrderOf` (P.getModuleName <$> sorted))
+  where
+    mkShallowModule P.ExternsFile{..} =
+      P.Module (P.internalModuleSourceSpan "<rebuild>") [] efModuleName (map mkImport efImports) Nothing
+    mkImport (P.ExternsImport mn it iq) =
+      P.ImportDeclaration mn it iq
+    getExtern = flip Map.lookup externs
+    -- Sort a list so its elements appear in the same order as in another list.
+    inOrderOf :: (Ord a) => [a] -> [a] -> [a]
+    inOrderOf xs ys = let s = Set.fromList xs in filter (`Set.member` s) ys
+
+------
+
+xrunSubsume
+  :: Environment
+  -> StateT TC.CheckState (SupplyT (WriterT b (Except e))) a
+  -> Either e (a, Environment)
+xrunSubsume env = runExcept . evalWriterT . P.evalSupplyT 0 . TC.runCheck' env
+
+evalWriterT :: Monad m => WriterT b m r -> m r
+evalWriterT m = liftM fst (runWriterT m)
+
+filtering
+  :: P.Environment
+  -- ^ The Environment which contains the relevant definitions and typeclasses
+  -> P.Type
+  -- ^ The type supplied by the environment
+  -> P.Type
+  -- ^ The user supplied type
+  -> Either P.MultipleErrors ((P.Expr, [(P.Ident, P.Constraint)]), P.Environment)
+filtering env x t = xrunSubsume env $ do
+  let initializeSkolems = Skolem.introduceSkolemScope <=< P.replaceAllTypeSynonyms <=< P.replaceTypeWildcards
+
+  x' <- initializeSkolems x
+  t' <- initializeSkolems t
+
+  let dummyExpression = P.Var (P.Qualified Nothing (P.Ident "x"))
+  elab <- runExceptT $ subsumes (Just dummyExpression) t' x'
+  subst <- gets TC.checkSubstitution
+  case elab of
+    Left _ -> throwError undefined
```

Comment:
I've tried this for about 30min now but can't figure it out. Dealing with transformers is still a little tricky for me. Maybe we could take a look at it together?